From b90e634cc2441ede9d30f0b7e52eb3e44c89efc0 Mon Sep 17 00:00:00 2001
From: Hongwei Wang <hwwang@google.com>
Date: Thu, 28 Aug 2025 11:19:07 -0700
Subject: [PATCH] GRANT TRUSTED_OVERLAY after PiP is entered

- Make sure the TRUSTED_OVERLAY is only granted after WMShell
  successfully handled the entering PiP transition
- Revoke TRUSTED_OVERLAY in WM once the task is not pinned any more

Flag: EXEMPT securityfix
Bug: 436270922
Test: atest PinnedStackTests
Merged-In: Id849d93bc92209fc296ee91ef70b0ffe2a09550e
Change-Id: Id849d93bc92209fc296ee91ef70b0ffe2a09550e
---
 .../src/com/android/wm/shell/pip/PipTransition.java  | 12 ++++++++++--
 .../com/android/server/wm/RootWindowContainer.java   |  9 ++++++---
 2 files changed, 16 insertions(+), 5 deletions(-)

diff --git a/libs/WindowManager/Shell/src/com/android/wm/shell/pip/PipTransition.java b/libs/WindowManager/Shell/src/com/android/wm/shell/pip/PipTransition.java
index e3d53fc415db..a206665b7e3e 100644
--- a/libs/WindowManager/Shell/src/com/android/wm/shell/pip/PipTransition.java
+++ b/libs/WindowManager/Shell/src/com/android/wm/shell/pip/PipTransition.java
@@ -380,16 +380,24 @@ public class PipTransition extends PipTransitionController {
             @PipAnimationController.TransitionDirection int direction,
             @NonNull SurfaceControl.Transaction tx) {
         final boolean enteringPip = isInPipDirection(direction);
+        final SurfaceControl leash = mPipOrganizer.getSurfaceControl();
+        final boolean hasValidLeash = leash != null && leash.isValid();
         if (enteringPip) {
             mPipTransitionState.setTransitionState(ENTERED_PIP);
+            // TRUSTED_OVERLAY is granted iff Shell successfully receives the transition.
+            // It's revoked once the task exits pinned mode in
+            // RootWindowContainer#notifyActivityPipModeChanged
+            if (hasValidLeash) {
+                ProtoLog.d(ShellProtoLogGroup.WM_SHELL_PICTURE_IN_PICTURE,
+                        "Set TRUSTED_OVERLAY for Task#%d", taskInfo.taskId);
+                tx.setTrustedOverlay(leash, true);
+            }
         }
         // If we have an exit transition, but aren't playing a transition locally, it
         // means we're expecting the exit transition will be "merged" into another transition
         // (likely a remote like launcher), so don't fire the finish-callback here -- wait until
         // the exit transition is merged.
         if ((mExitTransition == null || isAnimatingLocally()) && mFinishCallback != null) {
-            final SurfaceControl leash = mPipOrganizer.getSurfaceControl();
-            final boolean hasValidLeash = leash != null && leash.isValid();
             WindowContainerTransaction wct = null;
             if (isOutPipDirection(direction)) {
                 // Only need to reset surface properties. The server-side operations were already
diff --git a/services/core/java/com/android/server/wm/RootWindowContainer.java b/services/core/java/com/android/server/wm/RootWindowContainer.java
index 05f95f813e55..f5774009c287 100644
--- a/services/core/java/com/android/server/wm/RootWindowContainer.java
+++ b/services/core/java/com/android/server/wm/RootWindowContainer.java
@@ -2193,11 +2193,14 @@ class RootWindowContainer extends WindowContainer<DisplayContent>
             mService.getTaskChangeNotificationController().notifyActivityPinned(r);
         } else {
             mService.getTaskChangeNotificationController().notifyActivityUnpinned();
+            // Revoke the TRUSTED_OVERLAY here as a blanket policy.
+            if (task.getSurfaceControl() != null) {
+                mWmService.mTransactionFactory.get()
+                        .setTrustedOverlay(task.getSurfaceControl(), false)
+                        .apply();
+            }
         }
         mWindowManager.mPolicy.setPipVisibilityLw(inPip);
-        mWmService.mTransactionFactory.get()
-                .setTrustedOverlay(task.getSurfaceControl(), inPip)
-                .apply();
     }
 
     void executeAppTransitionForAllDisplay() {
-- 
2.52.0.rc1.455.g30608eb744-goog

